// liquid-node/lib/liquid/template.js
// Generated by CoffeeScript 1.7.1
(function() {
  var Liquid, Promise,
    __slice = [].slice,
    __hasProp = {}.hasOwnProperty;

  Liquid = require("../liquid");

  Promise = require("bluebird");

  module.exports = Liquid.Template = (function() {
    function Template() {
      this.registers = {};
      this.assigns = {};
      this.instanceAssigns = {};
      this.tags = {};
      this.errors = [];
      this.rethrowErrors = true;
    }

    Template.prototype.parse = function(engine, source) {
      this.engine = engine;
      if (source == null) {
        source = "";
      }
      return Promise["try"]((function(_this) {
        return function() {
          var tokens;
          tokens = _this._tokenize(source);
          _this.tags = _this.engine.tags;
          _this.root = new Liquid.Document(_this);
          return _this.root.parseWithCallbacks(tokens).then(function() {
            return _this;
          });
        };
      })(this));
    };

    Template.prototype.render = function() {
      var args;
      args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      return Promise["try"]((function(_this) {
        return function() {
          return _this._render.apply(_this, args);
        };
      })(this));
    };

    Template.prototype._render = function(assigns, options) {
      var context, k, v, _ref;
      if (this.root == null) {
        throw new Error("No document root. Did you parse the document yet?");
      }
      context = (function() {
        if (assigns instanceof Liquid.Context) {
          return assigns;
        } else if (assigns instanceof Object) {
          assigns = [assigns, this.assigns];
          return new Liquid.Context(this.engine, assigns, this.instanceAssigns, this.registers, this.rethrowErrors);
        } else if (assigns == null) {
          return new Liquid.Context(this.engine, this.assigns, this.instanceAssigns, this.registers, this.rethrowErrors);
        } else {
          throw new Error("Expected Object or Liquid::Context as parameter, but was " + (typeof assigns) + ".");
        }
      }).call(this);
      if (options != null ? options.registers : void 0) {
        _ref = options.registers;
        for (k in _ref) {
          if (!__hasProp.call(_ref, k)) continue;
          v = _ref[k];
          this.registers[k] = v;
        }
      }
      if (options != null ? options.filters : void 0) {
        context.registerFilters.apply(context, options.filters);
      }
      return this.root.render(context).then(function(chunks) {
        return Liquid.Helpers.toFlatString(chunks);
      })["finally"]((function(_this) {
        return function() {
          return _this.errors = context.errors;
        };
      })(this));
    };

    Template.prototype._tokenize = function(source) {
      var col, line, tokens;
      source = String(source);
      if (source.length === 0) {
        return [];
      }
      tokens = source.split(Liquid.TemplateParser);
      line = 1;
      col = 1;
      return tokens.filter(function(token) {
        return token.length > 0;
      }).map(function(value) {
        var lastIndex, linebreaks, result;
        result = {
          value: value,
          col: col,
          line: line
        };
        lastIndex = value.lastIndexOf("\n");
        if (lastIndex < 0) {
          col += value.length;
        } else {
          linebreaks = value.split("\n").length - 1;
          line += linebreaks;
          col = value.length - lastIndex;
        }
        return result;
      });
    };

    return Template;

  })();

}).call(this);
